### ISSUE TITLE
RewriteUnsafeSelect has bugs

### ISSUE THREAD

**Issue by liwchang** [dmlc/tvm#2000]
Two issues.
First, the rewriter might contradict with previous rewriting rules, since it is the last rewriter.
A simple example is using Pad and vectorize it, and codegen will trigger a llvm assertion.
RewriteUnsafeSelect happens after Vectorizer, but not consider Vectorizer's rewriting.

Second, UnsafeExprDetector in RewriteUnsafeSelect seems too conservative. 
Pad is still a good example. 
I saw TODO there, so it might be unfinished work.
A precise detector might be difficult.

IMO, RewriteUnsafeSelect is totally unnecessary, if semantics of select is well-defined in tvm's programming model. 

-L

**Comment by tqchen** [comment]
Thanks for bringing this up. Can you elaborate a bit on what is going on here? The second issue sounds like an overly conservative detection. 

Most normal user will use select as if_then_else and they are error-prone in the context of padding(depending on whether the device support OOB read, we want to be safe on it). So this rewriting is to support this semantics here. You can also think this as a rewriting if_then_else back to select.

**Comment by liwchang** [comment]
Yes, the second issue is about overly conservative detection.
The detector classified non-OOB to OOB. It might because  it failed to simplifying one branch of stmt. (The root reason I am not sure.)

In the current situation with Pad op test case, tvm rewrote select to if_then_else under vectorization, and failed llvm.
Fundamentally, it is not about OOB access. 
IMO, OOB shouldn't be involved in the programming model. 

It is more about semantics of select. 
IMO, the good semantics of select for general modern architectures will be something like this.

Both true and false stmt will executed before the condition stmt in arbitrary order. 
If there are exceptions or side-effect on those stmts, use as your risk.
(We could further say one stmt could be eliminated as one optimization. but that didn't impact the semantics)
This is similar to CUDA's definition for branch.
With this definition, users guard their safety for using select. 

Then, if a user has its own lower pass with particular hw support, and they can use unsafe select.
Then they can use unsafe one. 

Then, we can put that currently overly conservative UnsafeDetector to debug mode.
When you want to refine it, you refine later. :)

**Comment by tqchen** [comment]
I agree that this might be a better choice for performance. However, as select has also been extensively used for OOB read(padding), 

The main problem is most high-level languages don't have select at all and it is usually a low-level construct. But people tend to use select in the place which they actually want if_then_else. Given this situation, we take a compromise by using unsafe-select rewriter to rewrite the un-safe ones. 

We will likely need to keep this behavior for this compatibility reason. We can, however, introduce an new intrinsic e.g. branch which the unsafe rewriter ignores if necessary

**Comment by tqchen** [comment]
After thinking a bit more about this, I think I agree with you on pin down the semantics of Select. This, however, would involve an upgrade of most code since many users assumed the semantics of if_then_else when using select. So it is more of matter of a safe upgrade and when to do so, one possible way to is:

- introduce tvm.if_then_else, which serves as original select
- migrate the code to use if_then_else
- Keep the current UnsafeSelectDetector and throw an error to the user, to tell them to use if_then_else instead, this will at least give users a head up about this case
- allow to optionally turn the unsafe detector off

**Comment by liwchang** [comment]
Thanks, love you.
Let me confirm 1) Select will be kept. 
and we will have 2) capability of disabling RewriteUnsafeSelect.
This will work.

BTW, I need to correct one thing.
Pad I mentioned is the Pad op, topi's or defined in onnx. 
It doesn't rely on OOB read.
If i remember correctly in topi, you shift OOB read to edge read, so OOB won't happen.
